### PR TITLE
Model deserializer

### PR DESCRIPTION
**Pull request by AnastasiaTuchina**

### CODE REVIEW

File: serialization/src/main/kotlin/io/kinference/protobuf/message/AttributeProto.kt [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    val tensors: MutableList<TensorProto> = ArrayList(),
+    val graphs: MutableList<GraphProto> = ArrayList(),
+    val sparseTensors: MutableList<SparseTensorProto> = ArrayList()
+) {
+    companion object {
+        fun decode(reader: ProtobufReader): AttributeProto {
+            val proto = AttributeProto()
+            reader.forEachTag { tag ->
+                when (tag) {
+                    1 -> proto.name = reader.readString()
```
Comment (Fatalll) [inline serialization/src/main/kotlin/io/kinference/protobuf/message/AttributeProto.kt:29]: Can we use enum for tags?

File: serialization/src/main/kotlin/io/kinference/protobuf/message/AttributeProto.kt [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+                4 -> TENSOR
+                5 -> GRAPH
+                6 -> FLOATS
+                7 -> INTS
+                8 -> STRINGS
+                9 -> TENSORS
+                10 -> GRAPHS
+                11 -> SPARSE_TENSOR
+                12 -> SPARSE_TENSORS
+                else -> null
```
Comment (Fatalll) [inline serialization/src/main/kotlin/io/kinference/protobuf/message/AttributeProto.kt:92]: Do we really need nulls in types?

File: serialization/src/main/kotlin/io/kinference/protobuf/arrays/TiledArrayContainer.kt [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    private var tiledInitialized: Boolean = false
+
+    fun hasData() = array != null || tiled != null
+
+    fun setTiled(array: Any) {
+        tiled = array
+        tiledInitialized = true
+    }
+
+    fun decode(reader: ProtobufReader, tag: Int, dataType: TensorProto.DataType, shape: IntArray?) {
```
Comment (Fatalll) [inline serialization/src/main/kotlin/io/kinference/protobuf/arrays/TiledArrayContainer.kt:22]: In which cases we don't have shape?

File: serialization/src/main/kotlin/io/kinference/protobuf/arrays/DoubleArraySerializer.kt [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    override var position: Int = data.size
+        private set
+
+    init {
+        checkCapacity(INITIAL_CAPACITY)
+    }
+
+    override fun checkCapacity(requiredCapacity: Int) {
+        if (data.size < requiredCapacity)
+            data = data.copyOf(requiredCapacity.coerceAtLeast(data.size * 2))
```
Comment (Fatalll) [inline serialization/src/main/kotlin/io/kinference/protobuf/arrays/DoubleArraySerializer.kt:17]: Can we rewind the buffer to the next field and get array size based on the size of the skipped bytes?

File: serialization/src/main/kotlin/io/kinference/protobuf/message/TensorProto.kt [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+            }
+            if (rawData != null || !proto.hasData()) parseRaw(rawData, proto)
+            proto.checkTiledData()
+            return proto
+        }
+
+        private fun TensorProto.hasData() = _tiledData.hasData() || stringData.isNotEmpty() || externalData.isNotEmpty()
+
+        // convert data stored as int32 to the specified type
+        private fun TensorProto.checkTiledData() {
```
Comment (Fatalll) [inline serialization/src/main/kotlin/io/kinference/protobuf/message/TensorProto.kt:73]: Can we read int and immediately convert it to the desired type?

**Review by Fatalll — CHANGES_REQUESTED** [review]

File: inference/src/main/kotlin/io/kinference/operators/ml/trees/TreeEnsembleBuilder.kt [hunk 6]
```diff
@@ -104,7 +107,7 @@ internal class TreeEnsembleBuilder(private val info: TreeEnsembleOperator.BaseEn
 
         private fun parseLabelsInfo(info: TreeEnsembleClassifier.ClassifierInfo): TreeEnsemble.LabelsInfo {
             return when {
-                info.classlabels_int64s != null -> TreeEnsemble.LabelsInfo(info.classlabels_int64s!!, TensorProto.DataType.INT64)
+                info.classlabels_int64s != null -> TreeEnsemble.LabelsInfo(info.classlabels_int64s!!.toList(), TensorProto.DataType.INT64)
```
[the remark this answers is not in the document]
Comment (AnastasiaTuchina) [reply inference/src/main/kotlin/io/kinference/operators/ml/trees/TreeEnsembleBuilder.kt:110]: Label lists shouldn't be big (something like 10-20 values max), so it's ok

File: serialization/src/main/kotlin/io/kinference/protobuf/message/TensorProto.kt (hunk 5 above)
Comment (AnastasiaTuchina) [reply serialization/src/main/kotlin/io/kinference/protobuf/message/TensorProto.kt:73]: The data is not ordered, so it's more safe to read the full message and then process the data when we are guaranteed to have both the type and the data

File: serialization/src/main/kotlin/io/kinference/protobuf/arrays/DoubleArraySerializer.kt (hunk 4 above)
Comment (AnastasiaTuchina) [reply serialization/src/main/kotlin/io/kinference/protobuf/arrays/DoubleArraySerializer.kt:17]: No, wire protobuf reader doesn't allow that, but I tried:(

File: serialization/src/main/kotlin/io/kinference/protobuf/arrays/TiledArrayContainer.kt (hunk 3 above)
Comment (AnastasiaTuchina) [reply serialization/src/main/kotlin/io/kinference/protobuf/arrays/TiledArrayContainer.kt:22]: When we read the data, but shape wasn't initialized yet

**Review by Fatalll — APPROVED** [review]

**Pull request merged**